src/uucore/src/lib/features/checksum.rs

```diff
@@ -2,26 +2,967 @@
+            // Set default length to 512 if None
+            let bits = length.unwrap_or(512);
+            if bits == 512 {
+                Ok(HashAlgorithm {
+                    name: ALGORITHM_OPTIONS_BLAKE2B,
+                    create_fn: Box::new(move || Box::new(Blake2b::new())),
+                    bits: 512,
+                })
+            } else {
+                Ok(HashAlgorithm {
+                    name: ALGORITHM_OPTIONS_BLAKE2B,
+                    create_fn: Box::new(move || Box::new(Blake2b::with_output_bytes(bits))),
+                    bits,
+                })
+            }
+        }
+        ALGORITHM_OPTIONS_BLAKE3 | "b3sum" => Ok(HashAlgorithm {
+            name: ALGORITHM_OPTIONS_BLAKE3,
+            create_fn: Box::new(|| Box::new(Blake3::new())),
+            bits: 256,
+        }),
+        ALGORITHM_OPTIONS_SM3 => Ok(HashAlgorithm {
+            name: ALGORITHM_OPTIONS_SM3,
+            create_fn: Box::new(|| Box::new(Sm3::new())),
+            bits: 512,
+        }),
+        ALGORITHM_OPTIONS_SHAKE128 | "shake128sum" => {
+            let bits =
+                length.ok_or_else(|| USimpleError::new(1, "--bits required for SHAKE128"))?;
+            Ok(HashAlgorithm {
+                name: ALGORITHM_OPTIONS_SHAKE128,
+                create_fn: Box::new(|| Box::new(Shake128::new())),
+                bits,
+            })
+        }
+        ALGORITHM_OPTIONS_SHAKE256 | "shake256sum" => {
+            let bits =
+                length.ok_or_else(|| USimpleError::new(1, "--bits required for SHAKE256"))?;
+            Ok(HashAlgorithm {
+                name: ALGORITHM_OPTIONS_SHAKE256,
+                create_fn: Box::new(|| Box::new(Shake256::new())),
+                bits,
+            })
+        }
+        //ALGORITHM_OPTIONS_SHA3 | "sha3" => (
+        _ if algo.starts_with("sha3") => create_sha3(length),
+
+        _ => Err(ChecksumError::UnknownAlgorithm.into()),
+    }
+}
+
+// Regexp to handle the three input formats:
+// 1. <algo>[-<bits>] (<filename>) = <checksum>
+//    algo must be uppercase or b (for blake2b)
+// 2. <checksum> [* ]<filename>
+// 3. <checksum> [*]<filename> (only one space)
+const ALGO_BASED_REGEX: &str = r"^\s*\\?(?P<algo>(?:[A-Z0-9]+|BLAKE2b))(?:-(?P<bits>\d+))?\s?\((?P<filename>.*)\)\s*=\s*(?P<checksum>[a-fA-F0-9]+)$";
+const DOUBLE_SPACE_REGEX: &str = r"^(?P<checksum>[a-fA-F0-9]+)\s{2}(?P<filename>.*)$";
+
+// In this case, we ignore the *
+const SINGLE_SPACE_REGEX: &str = r"^(?P<checksum>[a-fA-F0-9]+)\s(?P<filename>\*?.*)$";
+
+fn get_filename_for_output(filename: &OsStr, input_is_stdin: bool) -> String {
+    if input_is_stdin {
+        "standard input"
+    } else {
+        filename.to_str().unwrap()
+    }
+    .maybe_quote()
+    .to_string()
+}
+
+/// Determines the appropriate regular expression to use based on the provided lines.
+fn determine_regex(
+    filename: &OsStr,
+    input_is_stdin: bool,
+    lines: &[String],
+) -> UResult<(Regex, bool)> {
```

Comment:
I'm not a fan of this return type and you already have to workaround it in the `perform_checksum_validation` function ;-) But I think fixing it is something for a future PR.